src/Sarif.ValidationTests/DefaultObjectPopulatingVisitor.cs

```diff
@@ -0,0 +1,293 @@
+            graphNodeParentCount++;
+            node = base.VisitNode(node);
+            graphNodeParentCount--;
+
+            return node;
+        }
+
+        private void PopulateInstanceWithDefaultMemberValues(ISarifNode node)        {
+           
+            Type nodeType = node.GetType();
+            var binding = BindingFlags.Public | BindingFlags.Instance;
+            foreach (PropertyInfo property in nodeType.GetProperties(binding))
+            {
+                // The node kind is always properly set in the OM and
+                // isn't relevant to the SARIF schema itself
+                if (property.Name == "SarifNodeKind") { continue; }
+
+                // Property bags and tags are populated via special methods on 
+                // the class rather than direct access of properties.
+                if (property.Name == "PropertyNames" || 
+                    property.Name == "Tags")
+                {
+                    continue;
+                }
+
+                PopulatePropertyWithDefaultValue(node, property);
+            }
+
+            // If we have a property bag holder (as most SARIF things are), we will 
+            // add and then immediately remove both a property and a tag. This has
+            // the effect of leaving non-null but empty properties collection.
+            var propertyBagHolder = node as PropertyBagHolder;
+            if (propertyBagHolder != null)
+            {
+                string meaninglessValue = "ToBeRemoved";
+                propertyBagHolder.SetProperty(propertyName: meaninglessValue, value: meaninglessValue);
+                propertyBagHolder.RemoveProperty(propertyName: meaninglessValue);
+
+                propertyBagHolder.Tags.Add(meaninglessValue);
+                propertyBagHolder.Tags.Remove(meaninglessValue);
+            }
+        }
+
+        private void PopulatePropertyWithDefaultValue(ISarifNode node, PropertyInfo property)
+        {
+            object propertyValue = null;
+            Type propertyType = property.PropertyType;
+
+            // If we can't set this property, it is not of interest
+            if (property.GetAccessors().Length != 2) { return; }
+
+            // This special-casing is required to account for the fact that an 
+            // exception instance itself contains an array of exceptions
+            // (exception.innerExceptions). We don't want to hydrate 
+            // the innerExceptions of any innerExceptions, which results
+            // in re-entrancy that ends up consuming all stack space.
+            //
+            // Once we have populated a single exceptions.innerExceptions array, 
+            // we have accomplished all the testing we need in any case.
+            if (node.SarifNodeKind == SarifNodeKind.ExceptionData &&
+                property.Name == "InnerExceptions" &&
+                exceptionDataParentCount > 0) { return; }
+
+            // Sample approach applies for graph node.children
```

Comment:
>Sample [](start = 15, length = 6)

"Sample" => "Similar"?

Reply:
Huh? How did autocorrect strike while I was typing this comment? ;)

---
In reply to: [229128421](https://github.com/Microsoft/sarif-sdk/pull/1121#discussion_r229128421) [](ancestors = 229128421)